File: lib/ch4/4.1.js
// Implement a function to check if a tree is balanced. For the purposes
// of this question, a balanced tree is defned to be a tree such that no
// two leaf nodes difer in distance from the root by more than one.

function isBalanced(node) {
  if (node === null) return true;

  const lHeight = height(node.left);
  const rHeight = height(node.right);

  return (
    Math.abs(lHeight - rHeight) <= 1
      && isBalanced(node.left)
      && isBalanced(node.right)
  );
}

function height(node) {
  if (node === null) return 0;
  return 1 + Math.max(height(node.left), height(node.right));
}

File: java/ch4/4.1.java
int depth(TreeNode root) {
    if (root == null) return 0;
    return Math.max(depth(root.left), depth(root.right)) + 1;
}
boolean isBalanced(TreeNode root) {
    if (root == null) return true;
    int leftDepth = depth(root.left);
    int rightDepth = depth(root.right);
    int diff = Math.abs(leftDepth - rightDepth);
    return (
        isBalanced(root.left) &&
        isBalanced(root.right) &&
        depth <= 1
    );
}
